File: route-persistence/src/androidTest/java/com/skedgo/routepersistence/RouteStoreTest.java
package com.skedgo.routepersistence;

import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import androidx.test.InstrumentationRegistry;
import androidx.test.espresso.core.deps.guava.collect.Lists;
import androidx.test.runner.AndroidJUnit4;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.skedgo.android.common.model.GsonAdaptersBooking;
import com.skedgo.android.common.model.GsonAdaptersRealtimeAlert;
import com.skedgo.android.common.model.Location;
import skedgo.tripkit.routing.Trip;
import skedgo.tripkit.routing.TripGroup;
import skedgo.tripkit.routing.TripSegment;
import com.skedgo.android.common.util.LowercaseEnumTypeAdapterFactory;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.UUID;
import java.util.concurrent.TimeUnit;

import rx.observers.TestSubscriber;

import static org.junit.Assert.assertEquals;

@RunWith(AndroidJUnit4.class)
public class RouteStoreTest {
  private RouteDatabaseHelper databaseHelper;
  private RouteStore store;

  @Before public void before() {
    final Context context = InstrumentationRegistry.getTargetContext();
    databaseHelper = new RouteDatabaseHelper(context, RouteStoreTest.class.getSimpleName());
    final Gson gson = new GsonBuilder()
        .registerTypeAdapterFactory(new LocationTypeAdapterFactory())
        .registerTypeAdapterFactory(new LowercaseEnumTypeAdapterFactory())
        .registerTypeAdapterFactory(new GsonAdaptersBooking())
        .registerTypeAdapterFactory(new GsonAdaptersRealtimeAlert())
        .create();
    store = new RouteStore(databaseHelper, gson);
  }

  @After public void after() {
    InstrumentationRegistry.getTargetContext().deleteDatabase(databaseHelper.getDatabaseName());
  }

  @Test public void shouldQueryExactSavedGroups() {
    final TripGroup group = new TripGroup();
    final Trip trip = new Trip();
    final TripSegment segment = new TripSegment();
    segment.setFrom(new Location(1.0, 2.0));
    segment.setTo(new Location(3.0, 4.0));
    trip.setSegments(Lists.newArrayList(segment));
    group.setTrips(Lists.newArrayList(trip));

    final String requestId = UUID.randomUUID().toString();
    store.saveAsync(requestId, Collections.singletonList(group))
        .toBlocking().first();

    final List<TripGroup> groups = store.queryAsync(GroupQueries.hasRequestId(requestId))
        .toList()
        .toBlocking().first();
    assertEquals(1, groups.size());
    final TripGroup actualGroup = groups.get(0);
    assertEquals(group.uuid(), actualGroup.uuid());

    final ArrayList<Trip> actualTrips = actualGroup.getTrips();
    assertEquals(1, actualTrips.size());

    final Trip actualTrip = actualTrips.get(0);
    assertEquals(1, actualTrip.getSegments().size());
  }

  @Test public void groupDeletionShouldTriggerTripDeletionAndSegmentDeletion() {
    final TripGroup group = new TripGroup();
    final Trip trip = new Trip();
    final TripSegment segment = new TripSegment();
    segment.setFrom(new Location(1.0, 2.0));
    segment.setTo(new Location(3.0, 4.0));
    trip.setSegments(Lists.newArrayList(segment));
    group.setTrips(Lists.newArrayList(trip));

    final String requestId = UUID.randomUUID().toString();
    store.saveAsync(requestId, Collections.singletonList(group))
        .toBlocking().first();
    store.deleteAsync(WhereClauses.matchesUuidOf(group))
        .toBlocking().first();

    // Deleting a group should trigger deleting its trips too.
    final SQLiteDatabase database = databaseHelper.getReadableDatabase();
    final Cursor tripCursor = database.rawQuery("select * from trips", null);
    assertEquals(0, tripCursor.getCount());
    tripCursor.close();

    // Deleting a trip should trigger deleting its segments.
    final Cursor segmentCursor = database.rawQuery("select * from segments", null);
    assertEquals(0, segmentCursor.getCount());
    segmentCursor.close();
  }

  @Test public void shouldQueryExactGroupHavingGivenUuid() {
    final String requestId = UUID.randomUUID().toString();
    final TripGroup group0 = new TripGroup();
    {
      final Trip trip = new Trip();
      final TripSegment segment = new TripSegment();
      segment.setFrom(new Location(1.0, 2.0));
      segment.setTo(new Location(3.0, 4.0));
      trip.setSegments(Lists.newArrayList(segment));
      group0.setTrips(Lists.newArrayList(trip));
    }
    store.saveAsync(requestId, Collections.singletonList(group0))
        .toBlocking().first();

    final TripGroup group1 = new TripGroup();
    {
      final Trip trip = new Trip();
      final TripSegment segment = new TripSegment();
      segment.setFrom(new Location(1.0, 2.0));
      segment.setTo(new Location(3.0, 4.0));
      trip.setSegments(Lists.newArrayList(segment));
      group0.setTrips(Lists.newArrayList(trip));
    }
    store.saveAsync(requestId, Collections.singletonList(group1))
        .toBlocking().first();

    final TestSubscriber<TripGroup> subscriber = new TestSubscriber<>();
    store.queryAsync(GroupQueries.hasUuid(group0.uuid())).subscribe(subscriber);

    subscriber.awaitTerminalEvent();
    subscriber.assertNoErrors();
    assertEquals(1, subscriber.getValueCount());

    final TripGroup group = subscriber.getOnNextEvents().get(0);
    assertEquals(group0.uuid(), group.uuid());
  }

  @Test public void shouldClearOldTripsWhenUpdatingGroup() {
    final String requestId = UUID.randomUUID().toString();
    final Trip trip0 = new Trip();
    final TripGroup group0 = new TripGroup();
    {
      final TripSegment segment = new TripSegment();
      segment.setFrom(new Location(1.0, 2.0));
      segment.setTo(new Location(3.0, 4.0));
      trip0.setSegments(Lists.newArrayList(segment));
      group0.setTrips(Lists.newArrayList(trip0));
    }
    store.saveAsync(requestId, Collections.singletonList(group0))
        .toBlocking().first();

    final Trip trip1 = new Trip();
    {
      final TripSegment segment = new TripSegment();
      segment.setFrom(new Location(1.0, 2.0));
      segment.setTo(new Location(3.0, 4.0));
      trip1.setSegments(Lists.newArrayList(segment));
      group0.setTrips(Lists.newArrayList(trip1));
    }
    store.saveAsync(requestId, Collections.singletonList(group0))
        .toBlocking().first();

    final TripGroup actualGroup = store.queryAsync(GroupQueries.hasUuid(group0.uuid()))
        .toBlocking().first();

    final ArrayList<Trip> actualTrips = actualGroup.getTrips();
    assertEquals(1, actualTrips.size());

    final Trip actualTrip = actualTrips.get(0);
    assertEquals(trip1.uuid(), actualTrip.uuid());
    assertEquals(1, actualTrip.getSegments().size());
  }

  @Test public void shouldDeletePastTripThatHappenedBefore2HoursAgo() {
    // Given a trip that happened before 2 hours ago.
    final Trip trip = new Trip();
    trip.setStartTimeInSecs(0L);
    trip.setEndTimeInSecs(TimeUnit.HOURS.toSeconds(1L));
    final TripGroup group = new TripGroup();
    {
      final TripSegment segment = new TripSegment();
      segment.setFrom(new Location(1.0, 2.0));
      segment.setTo(new Location(3.0, 4.0));
      trip.setSegments(Lists.newArrayList(segment));
      group.setTrips(Lists.newArrayList(trip));
    }
    final String requestId = UUID.randomUUID().toString();
    store.saveAsync(requestId, Collections.singletonList(group))
        .toBlocking().first();

    store.queryAsync(GroupQueries.hasUuid(group.uuid()))
        .toBlocking().first();

    final long currentMillis = TimeUnit.HOURS.toMillis(4L);
    final int deletedCount = store.deleteAsync(WhereClauses.happenedBefore(
        2, // hours
        currentMillis
    )).toBlocking().first();
    assertEquals(1, deletedCount);

    final TestSubscriber<TripGroup> subscriber = new TestSubscriber<>();
    store.queryAsync(GroupQueries.hasUuid(group.uuid()))
        .subscribe(subscriber);

    subscriber.awaitTerminalEvent();
    subscriber.assertNoValues();
  }

  @Test public void shouldNotDeletePastTripThatHappenedWithin2HoursAgo() {
    // Given a trip that happened within 2 hours ago.
    final Trip trip = new Trip();
    trip.setStartTimeInSecs(0L);
    trip.setEndTimeInSecs(TimeUnit.HOURS.toSeconds(3L));
    final TripGroup group = new TripGroup();
    {
      final TripSegment segment = new TripSegment();
      segment.setFrom(new Location(1.0, 2.0));
      segment.setTo(new Location(3.0, 4.0));
      trip.setSegments(Lists.newArrayList(segment));
      group.setTrips(Lists.newArrayList(trip));
    }
    final String requestId = UUID.randomUUID().toString();
    store.saveAsync(requestId, Collections.singletonList(group))
        .toBlocking().first();

    store.queryAsync(GroupQueries.hasUuid(group.uuid()))
        .toBlocking().first();

    final long currentMillis = TimeUnit.HOURS.toMillis(4L);
    final int deletedCount = store.deleteAsync(WhereClauses.happenedBefore(
        2, // hours
        currentMillis
    )).toBlocking().first();
    assertEquals(0, deletedCount);

    store.queryAsync(GroupQueries.hasUuid(group.uuid()))
        .toBlocking().first();
  }
}

File: route-persistence/src/test/java/com/skedgo/routepersistence/RouteStoreTest.kt
package com.skedgo.routepersistence

import android.content.Context
import androidx.test.core.app.ApplicationProvider
import com.google.gson.GsonBuilder
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.whenever
import com.skedgo.tripkit.common.model.GsonAdaptersBooking
import com.skedgo.tripkit.common.model.GsonAdaptersRealtimeAlert
import com.skedgo.tripkit.common.model.Location
import com.skedgo.tripkit.common.util.LowercaseEnumTypeAdapterFactory
import org.junit.Assert.assertEquals
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith

import androidx.test.ext.junit.runners.AndroidJUnit4
import com.skedgo.tripkit.routing.Trip
import com.skedgo.tripkit.routing.TripGroup
import com.skedgo.tripkit.routing.TripSegment
import java.util.*
import java.util.concurrent.TimeUnit

@RunWith(AndroidJUnit4::class)
class RouteStoreTest {
  private var databaseHelper: RouteDatabaseHelper? = null
  private lateinit var store: RouteStore

  @Before
  fun before() {
    val context = ApplicationProvider.getApplicationContext<Context>()
    databaseHelper = RouteDatabaseHelper(context, RouteStoreTest::class.java.simpleName)
    val gson = GsonBuilder()
        .registerTypeAdapterFactory(LocationTypeAdapterFactory())
        .registerTypeAdapterFactory(LowercaseEnumTypeAdapterFactory())
        .registerTypeAdapterFactory(GsonAdaptersBooking())
        .registerTypeAdapterFactory(GsonAdaptersRealtimeAlert())
        .create()
    store = RouteStore(databaseHelper!!, gson)
  }


  @Test
  fun shouldQueryExactSavedGroups() {
    val group = TripGroup()
    val trip = Trip()
    val segment = TripSegment()
    segment.from = Location(1.0, 2.0)
    segment.to = Location(3.0, 4.0)
    trip.segments = arrayListOf(segment)
    group.trips = arrayListOf(trip)

    val requestId = UUID.randomUUID().toString()
    store.saveAsync(requestId, listOf(group))
        .blockingFirst()

    val groups = store.queryAsync(GroupQueries.hasRequestId(requestId))
        .toList()
        .blockingGet()
    assertEquals(1, groups.size.toLong())
    val actualGroup = groups[0]
    assertEquals(group.uuid(), actualGroup.uuid())

    val actualTrips = actualGroup.trips
    assertEquals(1, actualTrips!!.size.toLong())

    val actualTrip = actualTrips[0]
    assertEquals(1, actualTrip.segments.size.toLong())
  }

  @Test
  fun groupDeletionShouldTriggerTripDeletionAndSegmentDeletion() {
    val group = TripGroup()
    val trip = Trip()
    val segment = TripSegment()
    segment.from = Location(1.0, 2.0)
    segment.to = Location(3.0, 4.0)
    trip.segments = arrayListOf(segment)
    group.trips = arrayListOf(trip)

    val requestId = UUID.randomUUID().toString()
    store.saveAsync(requestId, listOf(group))
        .blockingFirst()
    store.deleteAsync(WhereClauses.matchesUuidOf(group))
        .blockingFirst()

    // Deleting a group should trigger deleting its trips too.
    val database = databaseHelper!!.readableDatabase
    val tripCursor = database.rawQuery("select * from trips", null)
    assertEquals(0, tripCursor.count.toLong())
    tripCursor.close()

    // Deleting a trip should trigger deleting its segments.
    val segmentCursor = database.rawQuery("select * from segments", null)
    assertEquals(0, segmentCursor.count.toLong())
    segmentCursor.close()
  }

  @Test
  fun shouldQueryExactGroupHavingGivenUuid() {
    val requestId = UUID.randomUUID().toString()
    val group2 = TripGroup()
    run {
      val trip = Trip()
      val segment = TripSegment()
      segment.from = Location(1.0, 2.0)
      segment.to = Location(3.0, 4.0)
      trip.segments = arrayListOf(segment)
      group2.trips = arrayListOf(trip)
    }
    var observer = store.saveAsync(requestId, listOf(group2))
        .test()
    observer.awaitTerminalEvent()
    observer.assertComplete()

    val group1 = TripGroup()
    run {
      val trip = Trip()
      val segment = TripSegment()
      segment.from = Location(1.0, 2.0)
      segment.to = Location(3.0, 4.0)
      trip.segments = arrayListOf(segment)
      group1.trips = arrayListOf(trip)
    }
    observer = store.saveAsync(requestId, listOf(group1))
        .test()
    observer.awaitTerminalEvent()
    observer.assertComplete()

    val subscriber = store.queryAsync(GroupQueries.hasUuid(group2.uuid())).test()
    subscriber.awaitTerminalEvent()
    subscriber.assertNoErrors()
    assertEquals(1, subscriber.values().count().toLong())

    val group = subscriber.values().first()
    assertEquals(group2.uuid(), group.uuid())
  }

  @Test
  fun shouldClearOldTripsWhenUpdatingGroup() {
    val requestId = UUID.randomUUID().toString()
    val trip0 = Trip()
    val group0 = TripGroup()
    run {
      val segment = TripSegment()
      segment.from = Location(1.0, 2.0)
      segment.to = Location(3.0, 4.0)
      trip0.segments = arrayListOf(segment)
      group0.trips = arrayListOf(trip0)
    }
    store.saveAsync(requestId, listOf(group0))
        .blockingFirst()
    val trip1 = Trip()
    run {
      val segment = TripSegment()
      segment.from = Location(1.0, 2.0)
      segment.to = Location(3.0, 4.0)
      trip1.segments = arrayListOf(segment)
      group0.trips = arrayListOf(trip1)
    }
    store.saveAsync(requestId, listOf(group0))
        .blockingFirst()

    val actualGroup = store.queryAsync(GroupQueries.hasUuid(group0.uuid()))
        .blockingFirst()

    val actualTrips = actualGroup.trips
    assertEquals(1, actualTrips!!.size.toLong())

    val actualTrip = actualTrips[0]
    assertEquals(trip1.uuid(), actualTrip.uuid())
    assertEquals(1, actualTrip.segments.size.toLong())
  }

  @Test
  fun shouldDeletePastTripThatHappenedBefore2HoursAgo() {
    // Given a trip that happened before 2 hours ago.
    val trip = Trip()
    trip.startTimeInSecs = 0L
    trip.endTimeInSecs = TimeUnit.HOURS.toSeconds(1L)
    val group = TripGroup()
    run {
      val segment = TripSegment()
      segment.from = Location(1.0, 2.0)
      segment.to = Location(3.0, 4.0)
      trip.segments = arrayListOf(segment)
      group.trips = arrayListOf(trip)
    }
    val requestId = UUID.randomUUID().toString()
    store.saveAsync(requestId, listOf(group))
        .blockingFirst()

    store.queryAsync(GroupQueries.hasUuid(group.uuid()))
        .blockingFirst()

    val currentMillis = TimeUnit.HOURS.toMillis(4L)
    val deletedCount = store.deleteAsync(WhereClauses.happenedBefore(
        2, // hours
        currentMillis
    )).blockingFirst()
    assertEquals(1, deletedCount.toLong())

    val subscriber = store.queryAsync(GroupQueries.hasUuid(group.uuid())).test()
    subscriber.awaitTerminalEvent()
    subscriber.assertNoValues()
  }

  @Test
  fun `should add trip as display Trip`() {
    val trip = mock<Trip>().apply {
      whenever(this.uuid()).thenReturn("b")
      whenever(this.id).thenReturn(1)
    }
    val group = mock<TripGroup>().apply {
      whenever(this.uuid()).thenReturn("a")
      whenever(this.displayTrip).thenReturn(trip)
      whenever(this.trips).thenReturn(arrayListOf(trip))
    }
    group.displayTripId = 1
    group.trips = arrayListOf(trip)

    var observer = store.saveAsync(null, listOf(group))
        .test()
    observer.awaitTerminalEvent()
    observer.assertComplete()

    val newDisplayTrip = Trip().apply {
      this.uuid("c")
      this.id = 10
    }
    var observer2 = store.addTripToTripGroup("a", newDisplayTrip)
        .test()
    observer2.awaitTerminalEvent()
    observer2.assertComplete()

    var observer3 = store.queryAsync(GroupQueries.hasUuid("a"))
        .test()

    observer3.awaitTerminalEvent()
    observer3.assertComplete().assertValueCount(1)
        .values().first()
        .let {
          it.displayTripId `should equal` 2
          it.trips!!.map { it.uuid() } `should contain all` (listOf("b", "c"))
          it.displayTrip!!.uuid() `should equal` "c"
        }
  }

  @Test
  fun shouldNotDeletePastTripThatHappenedWithin2HoursAgo() {
    // Given a trip that happened within 2 hours ago.
    val trip = Trip()
    trip.startTimeInSecs = 0L
    trip.endTimeInSecs = TimeUnit.HOURS.toSeconds(3L)
    val group = TripGroup()
    run {
      val segment = TripSegment()
      segment.from = Location(1.0, 2.0)
      segment.to = Location(3.0, 4.0)
      trip.segments = arrayListOf(segment)
      group.trips = arrayListOf(trip)
    }
    val requestId = UUID.randomUUID().toString()
    store.saveAsync(requestId, listOf(group))
        .blockingFirst()

    store.queryAsync(GroupQueries.hasUuid(group.uuid()))
        .blockingFirst()

    val currentMillis = TimeUnit.HOURS.toMillis(4L)
    val deletedCount = store.deleteAsync(WhereClauses.happenedBefore(
        2, // hours
        currentMillis
    )).blockingFirst()
    assertEquals(0, deletedCount.toLong())

    store.queryAsync(GroupQueries.hasUuid(group.uuid()))
        .blockingFirst()
  }
}
